File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbact01c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of CardDemo batch program CBACT01C — Batch Account File Reader.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbact01c.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbact01c implements ProgramRunner {

    private static final BigDecimal BD_0_00 = new BigDecimal("0.00");
    private static final BigDecimal BD_2525_00 = new BigDecimal("2525.00");

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "PROCESS_RECORDS").toUpperCase();
        List<AccountRecord> accounts = buildScenario(scenarioName);
        if (accounts == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        // Process accounts
        List<String> displayLines = new ArrayList<>();
        List<OutAcctRec> outRecords = new ArrayList<>();
        List<ArrArrayRec> arrRecords = new ArrayList<>();
        List<VbRec1> vb1Records = new ArrayList<>();
        List<VbRec2> vb2Records = new ArrayList<>();
        int recordsRead = 0;

        for (AccountRecord acct : accounts) {
            recordsRead++;

            // 1100-DISPLAY-ACCT-RECORD
            List<String> lines = displayAcctRecord(acct);
            displayLines.addAll(lines);

            // 1300-POPUL + 1350-WRITE
            OutAcctRec outRec = buildOutRecord(acct);
            outRecords.add(outRec);

            // 1400-POPUL + 1450-WRITE
            arrRecords.add(buildArrRecord(acct));

            // 1500-POPUL + 1550/1575-WRITE
            VbRec1 vb1 = new VbRec1(acct.acctId, acct.acctActiveStatus);
            String reissueYyyy = acct.acctReissueDate.length() >= 4
                    ? acct.acctReissueDate.substring(0, 4) : "";
            VbRec2 vb2 = new VbRec2(acct.acctId, acct.acctCurrBal,
                    acct.acctCreditLimit, reissueYyyy);
            vb1Records.add(vb1);
            vb2Records.add(vb2);
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("RECORDS_READ", String.valueOf(recordsRead));
        out.put("OUT_RECORDS", String.valueOf(outRecords.size()));
        out.put("ARR_RECORDS", String.valueOf(arrRecords.size()));
        out.put("VB1_RECORDS", String.valueOf(vb1Records.size()));
        out.put("VB2_RECORDS", String.valueOf(vb2Records.size()));
        for (int i = 0; i < displayLines.size(); i++) {
            out.put("DISPLAY_" + i, displayLines.get(i));
        }
        for (int i = 0; i < outRecords.size(); i++) {
            out.put("OUT_CYC_DEBIT_" + i,
                    outRecords.get(i).acctCurrCycDebit.setScale(2, RoundingMode.HALF_UP).toPlainString());
        }
        return out;
    }

    private List<String> displayAcctRecord(AccountRecord acct) {
        List<String> lines = new ArrayList<>();
        lines.add("ACCT-ID                 :" + acct.acctId);
        lines.add("ACCT-ACTIVE-STATUS      :" + acct.acctActiveStatus);
        lines.add("ACCT-CURR-BAL           :" + acct.acctCurrBal);
        lines.add("ACCT-CREDIT-LIMIT       :" + acct.acctCreditLimit);
        lines.add("ACCT-CASH-CREDIT-LIMIT  :" + acct.acctCashCreditLimit);
        lines.add("ACCT-OPEN-DATE          :" + acct.acctOpenDate);
        lines.add("ACCT-EXPIRAION-DATE     :" + acct.acctExpirationDate);
        lines.add("ACCT-REISSUE-DATE       :" + acct.acctReissueDate);
        lines.add("ACCT-CURR-CYC-CREDIT    :" + acct.acctCurrCycCredit);
        lines.add("ACCT-CURR-CYC-DEBIT     :" + acct.acctCurrCycDebit);
        lines.add("ACCT-GROUP-ID           :" + acct.acctGroupId);
        // 49 dashes
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 49; i++) sb.append('-');
        lines.add(sb.toString());
        return lines;
    }

    private OutAcctRec buildOutRecord(AccountRecord acct) {
        // Reformat date (trivially, same format in/out for type "2"->"2")
        String reissueReformatted = acct.acctReissueDate;

        BigDecimal cycDebit = acct.acctCurrCycDebit;
        if (cycDebit.compareTo(BigDecimal.ZERO) == 0) {
            cycDebit = BD_2525_00;
        }

        return new OutAcctRec(
                acct.acctId, acct.acctActiveStatus, acct.acctCurrBal,
                acct.acctCreditLimit, acct.acctCashCreditLimit,
                acct.acctOpenDate, acct.acctExpirationDate,
                reissueReformatted, acct.acctCurrCycCredit, cycDebit,
                acct.acctGroupId
        );
    }

    private ArrArrayRec buildArrRecord(AccountRecord acct) {
        BigDecimal[][] balances = new BigDecimal[5][2];
        for (int i = 0; i < 5; i++) {
            balances[i][0] = BD_0_00;
            balances[i][1] = BD_0_00;
        }
        balances[0][0] = acct.acctCurrBal;
        balances[0][1] = new BigDecimal("1005.00");
        balances[1][0] = acct.acctCurrBal;
        balances[1][1] = new BigDecimal("1525.00");
        balances[2][0] = new BigDecimal("-1025.00");
        balances[2][1] = new BigDecimal("-2500.00");
        return new ArrArrayRec(acct.acctId, balances);
    }

    // ── Scenarios ─────────────────────────────────────────────────────────

    private List<AccountRecord> buildScenario(String name) {
        switch (name) {
            case "PROCESS_RECORDS": return scenarioProcessRecords();
            case "EMPTY_INPUT":    return scenarioEmptyInput();
            default: return null;
        }
    }

    private List<AccountRecord> scenarioProcessRecords() {
        List<AccountRecord> list = new ArrayList<>();
        list.add(new AccountRecord(
                100000001, "Y", new BigDecimal("5000.00"),
                new BigDecimal("10000.00"), new BigDecimal("3000.00"),
                "2020-03-15", "2029-12-31", "2025-03-15",
                new BigDecimal("1500.00"), BD_0_00, "GOLD"
        ));
        list.add(new AccountRecord(
                100000002, "Y", new BigDecimal("2500.50"),
                new BigDecimal("5000.00"), new BigDecimal("1500.00"),
                "2021-07-01", "2028-06-30", "2024-07-01",
                new BigDecimal("800.00"), new BigDecimal("200.00"), "PLAT"
        ));
        return list;
    }

    private List<AccountRecord> scenarioEmptyInput() {
        return new ArrayList<>();
    }

    // ── Inner data types ─────────────────────────────────────────────────

    static class AccountRecord {
        final int acctId;
        final String acctActiveStatus;
        final BigDecimal acctCurrBal;
        final BigDecimal acctCreditLimit;
        final BigDecimal acctCashCreditLimit;
        final String acctOpenDate;
        final String acctExpirationDate;
        final String acctReissueDate;
        final BigDecimal acctCurrCycCredit;
        final BigDecimal acctCurrCycDebit;
        final String acctGroupId;

        AccountRecord(int acctId, String activeStatus, BigDecimal currBal,
                      BigDecimal creditLimit, BigDecimal cashCreditLimit,
                      String openDate, String expirationDate, String reissueDate,
                      BigDecimal currCycCredit, BigDecimal currCycDebit,
                      String groupId) {
            this.acctId = acctId;
            this.acctActiveStatus = activeStatus;
            this.acctCurrBal = currBal;
            this.acctCreditLimit = creditLimit;
            this.acctCashCreditLimit = cashCreditLimit;
            this.acctOpenDate = openDate;
            this.acctExpirationDate = expirationDate;
            this.acctReissueDate = reissueDate;
            this.acctCurrCycCredit = currCycCredit;
            this.acctCurrCycDebit = currCycDebit;
            this.acctGroupId = groupId;
        }
    }

    static class OutAcctRec {
        final int acctId;
        final String acctActiveStatus;
        final BigDecimal acctCurrBal;
        final BigDecimal acctCreditLimit;
        final BigDecimal acctCashCreditLimit;
        final String acctOpenDate;
        final String acctExpirationDate;
        final String acctReissueDate;
        final BigDecimal acctCurrCycCredit;
        final BigDecimal acctCurrCycDebit;
        final String acctGroupId;

        OutAcctRec(int acctId, String activeStatus, BigDecimal currBal,
                   BigDecimal creditLimit, BigDecimal cashCreditLimit,
                   String openDate, String expirationDate, String reissueDate,
                   BigDecimal currCycCredit, BigDecimal currCycDebit,
                   String groupId) {
            this.acctId = acctId;
            this.acctActiveStatus = activeStatus;
            this.acctCurrBal = currBal;
            this.acctCreditLimit = creditLimit;
            this.acctCashCreditLimit = cashCreditLimit;
            this.acctOpenDate = openDate;
            this.acctExpirationDate = expirationDate;
            this.acctReissueDate = reissueDate;
            this.acctCurrCycCredit = currCycCredit;
            this.acctCurrCycDebit = currCycDebit;
            this.acctGroupId = groupId;
        }
    }

    static class ArrArrayRec {
        final int acctId;
        final BigDecimal[][] balances;

        ArrArrayRec(int acctId, BigDecimal[][] balances) {
            this.acctId = acctId;
            this.balances = balances;
        }
    }

    static class VbRec1 {
        final int acctId;
        final String acctActiveStatus;

        VbRec1(int acctId, String activeStatus) {
            this.acctId = acctId;
            this.acctActiveStatus = activeStatus;
        }
    }

    static class VbRec2 {
        final int acctId;
        final BigDecimal acctCurrBal;
        final BigDecimal acctCreditLimit;
        final String acctReissueYyyy;

        VbRec2(int acctId, BigDecimal currBal, BigDecimal creditLimit,
               String reissueYyyy) {
            this.acctId = acctId;
            this.acctCurrBal = currBal;
            this.acctCreditLimit = creditLimit;
            this.acctReissueYyyy = reissueYyyy;
        }
    }
}

File: pipeline/reimpl/python/cbact01c.py
"""
Reimplementation of CBACT01C — CardDemo Batch Account File Reader.

Function: Read the account VSAM KSDS sequentially and write records to:
  - OUT-FILE:   flat sequential file with reformatted fields
  - ARRY-FILE:  array record (account ID + balance array)
  - VBRC-FILE:  variable-length records (VB1=12 bytes, VB2=39 bytes)

Also calls COBDATFT for date reformatting (replicated inline here).
"""

from __future__ import annotations
from dataclasses import dataclass, field
from decimal import Decimal
from datetime import datetime
from typing import Iterator, Callable

from .carddemo_data import AccountRecord, get_db2_timestamp


# ── Output record structures ──────────────────────────────────────────────────

@dataclass
class OutAcctRec:
    """OUT-FILE record — fixed sequential output."""
    acct_id: int = 0
    acct_active_status: str = ""
    acct_curr_bal: Decimal = Decimal("0.00")
    acct_credit_limit: Decimal = Decimal("0.00")
    acct_cash_credit_limit: Decimal = Decimal("0.00")
    acct_open_date: str = ""
    acct_expiration_date: str = ""
    acct_reissue_date: str = ""
    acct_curr_cyc_credit: Decimal = Decimal("0.00")
    acct_curr_cyc_debit: Decimal = Decimal("0.00")
    acct_group_id: str = ""


@dataclass
class ArrAcctBal:
    curr_bal: Decimal = Decimal("0.00")
    curr_cyc_debit: Decimal = Decimal("0.00")


@dataclass
class ArrArrayRec:
    """ARRY-FILE record — array of 5 balance entries."""
    acct_id: int = 0
    balances: list[ArrAcctBal] = field(
        default_factory=lambda: [ArrAcctBal() for _ in range(5)]
    )


@dataclass
class VbRec1:
    """Variable-length record type 1 (12 bytes)."""
    acct_id: int = 0
    acct_active_status: str = ""


@dataclass
class VbRec2:
    """Variable-length record type 2 (39 bytes)."""
    acct_id: int = 0
    acct_curr_bal: Decimal = Decimal("0.00")
    acct_credit_limit: Decimal = Decimal("0.00")
    acct_reissue_yyyy: str = ""


@dataclass
class ProcessResult:
    out_records: list[OutAcctRec] = field(default_factory=list)
    arr_records: list[ArrArrayRec] = field(default_factory=list)
    vb1_records: list[VbRec1] = field(default_factory=list)
    vb2_records: list[VbRec2] = field(default_factory=list)
    records_read: int = 0
    display_lines: list[str] = field(default_factory=list)


# ── Date conversion (inlined from COBDATFT) ───────────────────────────────────

def _reformat_date(date_str: str, in_type: str, out_type: str) -> str:
    """Simplified COBDATFT date reformatter.

    type '2' = YYYY-MM-DD (ISO),  type '1' = MM/DD/YYYY
    """
    if not date_str or date_str.strip() == "":
        return date_str
    try:
        if in_type == "2":
            # Input: YYYY-MM-DD
            y, m, d = date_str.strip().split("-")
        elif in_type == "1":
            m, d, y = date_str.strip().split("/")
        else:
            return date_str

        if out_type == "2":
            return f"{y}-{m}-{d}"
        elif out_type == "1":
            return f"{m}/{d}/{y}"
        return date_str
    except (ValueError, AttributeError):
        return date_str


# ── Core processing ───────────────────────────────────────────────────────────

def process_account_file(
    accounts: list[AccountRecord],
    logger: Callable[[str], None] = print,
) -> ProcessResult:
    """Process all account records — mirrors CBACT01C PROCEDURE DIVISION.

    For each account:
      1. Display the account record (1100-DISPLAY-ACCT-RECORD)
      2. Build and write OUT-FILE record (1300-POPUL / 1350-WRITE)
      3. Build and write ARRY-FILE record (1400-POPUL / 1450-WRITE)
      4. Build and write VBRC records (1500-POPUL / 1550,1575-WRITE)
    """
    result = ProcessResult()
    logger("START OF EXECUTION OF PROGRAM CBACT01C")

    for acct in accounts:
        result.records_read += 1

        # 1100-DISPLAY-ACCT-RECORD
        lines = _display_acct_record(acct)
        for line in lines:
            logger(line)
        result.display_lines.extend(lines)

        # 1300-POPUL-ACCT-RECORD + 1350-WRITE
        out_rec = _build_out_record(acct)
        result.out_records.append(out_rec)

        # 1400-POPUL-ARRAY-RECORD + 1450-WRITE
        arr_rec = _build_arr_record(acct)
        result.arr_records.append(arr_rec)

        # 1500-POPUL-VBRC-RECORD + 1550/1575-WRITE
        vb1, vb2 = _build_vb_records(acct)
        logger(f"VBRC-REC1:{vb1}")
        logger(f"VBRC-REC2:{vb2}")
        result.vb1_records.append(vb1)
        result.vb2_records.append(vb2)

    logger("END OF EXECUTION OF PROGRAM CBACT01C")
    return result


def _display_acct_record(acct: AccountRecord) -> list[str]:
    return [
        f"ACCT-ID                 :{acct.acct_id}",
        f"ACCT-ACTIVE-STATUS      :{acct.acct_active_status}",
        f"ACCT-CURR-BAL           :{acct.acct_curr_bal}",
        f"ACCT-CREDIT-LIMIT       :{acct.acct_credit_limit}",
        f"ACCT-CASH-CREDIT-LIMIT  :{acct.acct_cash_credit_limit}",
        f"ACCT-OPEN-DATE          :{acct.acct_open_date}",
        f"ACCT-EXPIRAION-DATE     :{acct.acct_expiration_date}",
        f"ACCT-REISSUE-DATE       :{acct.acct_reissue_date}",
        f"ACCT-CURR-CYC-CREDIT    :{acct.acct_curr_cyc_credit}",
        f"ACCT-CURR-CYC-DEBIT     :{acct.acct_curr_cyc_debit}",
        f"ACCT-GROUP-ID           :{acct.acct_group_id}",
        "-" * 49,
    ]


def _build_out_record(acct: AccountRecord) -> OutAcctRec:
    """1300-POPUL-ACCT-RECORD logic."""
    # MOVE ACCT-REISSUE-DATE TO CODATECN-INP-DATE; CALL 'COBDATFT'
    reissue_reformatted = _reformat_date(acct.acct_reissue_date, "2", "2")

    cyc_debit = acct.acct_curr_cyc_debit
    if cyc_debit == Decimal("0"):
        cyc_debit = Decimal("2525.00")  # COBOL: IF = 0 MOVE 2525.00

    return OutAcctRec(
        acct_id=acct.acct_id,
        acct_active_status=acct.acct_active_status,
        acct_curr_bal=acct.acct_curr_bal,
        acct_credit_limit=acct.acct_credit_limit,
        acct_cash_credit_limit=acct.acct_cash_credit_limit,
        acct_open_date=acct.acct_open_date,
        acct_expiration_date=acct.acct_expiration_date,
        acct_reissue_date=reissue_reformatted,
        acct_curr_cyc_credit=acct.acct_curr_cyc_credit,
        acct_curr_cyc_debit=cyc_debit,
        acct_group_id=acct.acct_group_id,
    )


def _build_arr_record(acct: AccountRecord) -> ArrArrayRec:
    """1400-POPUL-ARRAY-RECORD logic.

    ARR-ACCT-CURR-BAL(1) = ACCT-CURR-BAL, DEBIT(1) = 1005.00
    ARR-ACCT-CURR-BAL(2) = ACCT-CURR-BAL, DEBIT(2) = 1525.00
    ARR-ACCT-CURR-BAL(3) = -1025.00,      DEBIT(3) = -2500.00
    Elements 4 and 5 remain zero (INITIALIZE)
    """
    balances = [ArrAcctBal() for _ in range(5)]
    balances[0] = ArrAcctBal(acct.acct_curr_bal, Decimal("1005.00"))
    balances[1] = ArrAcctBal(acct.acct_curr_bal, Decimal("1525.00"))
    balances[2] = ArrAcctBal(Decimal("-1025.00"), Decimal("-2500.00"))
    return ArrArrayRec(acct_id=acct.acct_id, balances=balances)


def _build_vb_records(acct: AccountRecord) -> tuple[VbRec1, VbRec2]:
    """1500-POPUL-VBRC-RECORD logic."""
    vb1 = VbRec1(
        acct_id=acct.acct_id,
        acct_active_status=acct.acct_active_status,
    )
    # VB2: reissue year extracted from YYYY-MM-DD
    reissue_yyyy = acct.acct_reissue_date[:4] if acct.acct_reissue_date else ""
    vb2 = VbRec2(
        acct_id=acct.acct_id,
        acct_curr_bal=acct.acct_curr_bal,
        acct_credit_limit=acct.acct_credit_limit,
        acct_reissue_yyyy=reissue_yyyy,
    )
    return vb1, vb2


# ── run_vector adapter ───────────────────────────────────────────────────────

def _scenario_process_records():
    return [
        AccountRecord(
            acct_id=100000001,
            acct_active_status="Y",
            acct_curr_bal=Decimal("5000.00"),
            acct_credit_limit=Decimal("10000.00"),
            acct_cash_credit_limit=Decimal("3000.00"),
            acct_open_date="2020-03-15",
            acct_expiration_date="2029-12-31",
            acct_reissue_date="2025-03-15",
            acct_curr_cyc_credit=Decimal("1500.00"),
            acct_curr_cyc_debit=Decimal("0.00"),
            acct_group_id="GOLD",
        ),
        AccountRecord(
            acct_id=100000002,
            acct_active_status="Y",
            acct_curr_bal=Decimal("2500.50"),
            acct_credit_limit=Decimal("5000.00"),
            acct_cash_credit_limit=Decimal("1500.00"),
            acct_open_date="2021-07-01",
            acct_expiration_date="2028-06-30",
            acct_reissue_date="2024-07-01",
            acct_curr_cyc_credit=Decimal("800.00"),
            acct_curr_cyc_debit=Decimal("200.00"),
            acct_group_id="PLAT",
        ),
    ]


def _scenario_empty_input():
    return []


_SCENARIOS = {
    "PROCESS_RECORDS": _scenario_process_records,
    "EMPTY_INPUT": _scenario_empty_input,
}


def run_vector(inputs: dict) -> dict:
    """Adapter for the differential harness runner contract."""
    scenario_name = str(inputs.get("SCENARIO", "PROCESS_RECORDS")).upper()
    if scenario_name not in _SCENARIOS:
        return {"error": f"unknown scenario: {scenario_name!r}"}

    accounts = _SCENARIOS[scenario_name]()
    result = process_account_file(accounts, logger=lambda _: None)

    out: dict[str, str] = {
        "RECORDS_READ": str(result.records_read),
        "OUT_RECORDS": str(len(result.out_records)),
        "ARR_RECORDS": str(len(result.arr_records)),
        "VB1_RECORDS": str(len(result.vb1_records)),
        "VB2_RECORDS": str(len(result.vb2_records)),
    }
    for i, line in enumerate(result.display_lines):
        out[f"DISPLAY_{i}"] = line
    for i, orec in enumerate(result.out_records):
        cyc_debit_str = f"{orec.acct_curr_cyc_debit:.2f}"
        out[f"OUT_CYC_DEBIT_{i}"] = cyc_debit_str
    return out
